### ISSUE TITLE
build broken for Intel 17 compilers

### ISSUE THREAD

**Issue by khuck** [STEllAR-GROUP/hpx#2904]
see: http://rostam.cct.lsu.edu/builders/hpx_intel_17_2_174_boost_1_59_centos_x86_64_debug/builds/16/steps/build_core/logs/stdio

**Comment by K-ballo** [comment]
The same errors can be observed on clang when building the branch `pack-short-circuit`. They seem to come from a poisonous overload in the local/remote `dataflow` set, and are likely related to the recent work on pack traversal.

**Comment by Naios** [comment]
Pack traversal isn't used by `dataflow` since #2829 isn't merged yet.

**Comment by K-ballo** [comment]
In that case then there's a tiny chance that the work on pack traversal fixes it, as the error seems to come from a poisonous computation of a fused invocation. I'll apply `pack-short-circuit` to that PR and see what happens...

**Comment by K-ballo** [comment]
Unfortunately applying that PR does not solve the issue.

@Naios You can see from the logs that the poisonous overload is caused by a bad interaction with `util::detail::functional_unwrap_impl`, which I had assumed to be part of the pack traversal related work. There seems to be some sort of disagreement between what `dataflow` and `unwrap` think is callable, and to make it worse the issue seems to be instantiation order sensitive.

For completeness, here's the same failure as experienced with clang (on the `pack-short-circuit` branch): https://circleci.com/gh/STEllAR-GROUP/hpx/8033

**Comment by K-ballo** [comment]
I've reduced the clang/pack-short-circuit test case to this:

    #include <hpx/lcos/future.hpp>
    #include <hpx/util/unwrap.hpp>

    void f(int) {}
    int main() {
        hpx::util::unwrapping(f)
            (hpx::lcos::make_ready_future(42), hpx::lcos::make_ready_future());
    }

Does this also fail on Intel 17 for the master branch?

**Comment by Naios** [comment]
@K-ballo is this test failing on clang/MSVC as well?

**Comment by Naios** [comment]
Did you test whether

```cpp
#include <hpx/lcos/future.hpp>
#include <hpx/util/unwrap.hpp>

std::true_type tt =
  hpx::util::detail::is_functional_unwrap<
    decltype(hpx::lcos::make_ready_future(42)),
    decltype(hpx::lcos::make_ready_future())
  >{};
```

is still valid after applying the short circuit branch?
So there is no issue in the sort circuit implementation itself.
I took a short look over the branch and your improvements seem fine for me.

**Comment by K-ballo** [comment]
That snippet isn't valid neither after nor before. I fail to see how this is related, since the header that defines the trait isn't even included (including it doesn't make the code valid either).

In any event, I suggest focusing on fixing the issue with `unwrapping` on the master branch.

**Comment by K-ballo** [comment]
This particular issue has been determined to be caused by pack traversal failing to map ranges of future types for Intel (most likely a bug on their end); more specifically, by the mechanism for remapping a container's allocator.

See: https://github.com/STEllAR-GROUP/hpx/tree/fixing-2904 for the minimum changes that "fix" the core build.

**Comment by hkaiser** [comment]
This was fixed by merging #2981.
